### DISCUSSION TITLE
Need help with select tag construction using form helpers

### DISCUSSION THREAD

**Discussion by cfcoder** [wheels-dev/wheels#1383] [Q&A] [answered]
I've looked at the docs for the form helper for select.  It's not clear why there is a select() and selectTag() but that's not my primary question. How would I convert the following code into a form helper using select()?

```
<select name="entry_type">
  <option <cfif journalEntry.Entry_Type eq "Accomplishment"> selected </cfif> value="Accomplishment">Accomplishment</option>
  <option <cfif journalEntry.Entry_Type eq "Comment"> selected </cfif> value="Comments">Comment</option>
  <option <cfif journalEntry.Entry_Type eq "Scope Adjustment"> selected </cfif> value="Scope Adjustment">Scope Adjustment</option>
</select>
```
Notice that it selects the existing value for editing an existing record. Also notice that the value stored in the db is the same thing as what is shown in the UI.  It's not obvious how to do this.  I tried creating a struct like this:

```
<cfscript>
	entryTypeOptions = {};
	entryTypeOptions["1"] = "Accomplishment";
	entryTypeOptions["2"] = "Comment";
	entryTypeOptions["3"] = "Scope Adjustment";
</cfscript> 
#select(property="entry_type", objectName="journalEntry", options="#entryTypeOptions#")#
```
But it doesn't pre-select existing values. It also stores the number value or key of the struct. I'm looking for clean and simple here. Maybe there isn't?

**Comment by cfcoder** [comment]
It also appears to not work to mix form helpers with regular html tags either such as this:

```
<select name="entry_type" property="entry_type" objectName="journalEntry">
	<option <cfif journalEntry.Entry_Type eq "Accomplishment"> selected </cfif> value="Accomplishment">Accomplishment</option>
	<option <cfif journalEntry.Entry_Type eq "Comment"> selected </cfif> value="Comments">Comment</option>
	<option <cfif journalEntry.Entry_Type eq "Scope Adjustment"> selected </cfif> value="Scope Adjustment">Scope Adjustment</option>
</select>
```
As this submitted value shows up in params scope and not in the params.journalEntry object. :(
Is that just a no-no... mixing helper methods with html tags inside a startFormTag() wrapper?

**Comment by cfcoder** [comment, marked as answer]
Just found that when I use an array of string values for the options, it all just works fine.
`entryTypeOptions = ["Accomplishment", "Comment", "Scope Adjustment"];`

Because the objectName object has an entry_type element with value that matches one of those in that array, it does pre-select it and then if changed and the form is submitted, it captures and stores the new value chosen in the params.journalEntry.entry_type.

**Comment by chapmandu** [comment]
FYI, `select(objectName = "foo")` is only for use with model instances, `select(name = "foo")` is a general HTML select list.

With `select()` `options` argument you can provide a query, list of strings, an array of strings, or an array of structs.

**Reply by cfcoder** [reply]
Using a struct didn't work so well for me. The array works.

**Reply by chapmandu** [reply]
My go to data format for select options is an array of structs:
`[{"value": 666, "text": "Foo"}]`

**Reply by cfcoder** [reply]
To get the pre-selected option to show, does that require setting the valueField and textField attributes?

**Reply by chapmandu** [reply]
I don't think you need to use those arguments with `select()`. If the instance.myProperty equals one of the option values, then it should be selected.
